crates/holochain_sqlite/src/db/p2p_agent_store.rs

```diff
@@ -113,6 +113,177 @@ impl AsP2pAgentStoreConExt for crate::db::PConnGuard {
+        let _ = self.0.lock().remove(agent);
+        Ok(())
+    }
+
+    fn get(&self, agent: &KitsuneAgent) -> DatabaseResult<Option<AgentInfoSigned>> {
+        Ok(self.0.lock().get(agent).cloned())
+    }
+
+    fn get_all(&self) -> DatabaseResult<Vec<AgentInfoSigned>> {
+        Ok(self
+            .0
+            .lock()
+            .values()
+            .filter_map(|info| {
+                if !info.is_active() {
+                    return None;
+                }
+
+                Some(info.clone())
+            })
+            .collect())
+    }
+
+    fn count(&self) -> DatabaseResult<u32> {
+        Ok(self.0.lock().len() as u32)
+    }
+
+    fn query_agents(
+        &self,
+        since_ms: u64,
+        until_ms: u64,
+        arcset: DhtArcSet,
+    ) -> DatabaseResult<Vec<AgentInfoSigned>> {
+        Ok(self
+            .0
+            .lock()
+            .values()
+            .filter_map(|info| {
+                if !info.is_active() {
+                    return None;
+                }
+
+                if info.signed_at_ms < since_ms {
+                    return None;
+                }
+
+                if info.signed_at_ms > until_ms {
+                    return None;
+                }
+
+                let interval = DhtArcRange::from(info.storage_arc);
+                if !arcset.overlap(&interval.into()) {
+                    return None;
+                }
+
+                Some(info.clone())
+            })
+            .collect())
+    }
+
+    fn query_near_basis(&self, basis: u32, limit: u32) -> DatabaseResult<Vec<AgentInfoSigned>> {
+        let lock = self.0.lock();
+
+        let mut out: Vec<(u32, &AgentInfoSigned)> = lock
+            .values()
+            .filter_map(|v| {
+                if v.is_active() {
+                    Some((v.storage_arc.dist(basis), v))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if out.len() > 1 {
+            out.sort_by(|a, b| a.0.cmp(&b.0));
+        }
+
+        Ok(out
+            .into_iter()
+            .take(limit as usize)
+            .map(|(_, v)| v.clone())
+            .collect())
+    }
+}
+
+struct AgentStoreByPath {
+    map: Mutex<HashMap<String, Arc<AgentStore>>>,
+}
+
+impl AgentStoreByPath {
+    fn new() -> Self {
+        Self {
+            map: Mutex::new(HashMap::new()),
+        }
+    }
+
+    fn get(&self, con: &Connection) -> DatabaseResult<Arc<AgentStore>> {
+        match self.map.lock().entry(con.path().unwrap().to_string()) {
+            hash_map::Entry::Occupied(e) => Ok(e.get().clone()),
+            hash_map::Entry::Vacant(e) => {
+                let agent_store = Arc::new(AgentStore::new(con)?);
+                e.insert(agent_store.clone());
+                Ok(agent_store)
+            }
+        }
+    }
+}
+
+static AI_CACHE: Lazy<AgentStoreByPath> = Lazy::new(AgentStoreByPath::new);
```

Comment:
AI is an unfortunate acronym here, took me a moment to figure out what it meant. I vote `AGENT_CACHE` but I'm almost completely indifferent.

Reply:
ya... okay, I'll change that : )

Reply:
Done. Went with just `CACHE` for the static and `cache_get` for the function, since we're already in the p2p_agent_store module.